applications/FluidDynamicsApplication/custom_utilities/estimate_dt_utilities.cpp

```diff
@@ -64,28 +93,141 @@ namespace Kratos
+        const auto& r_geom = mrModelPart.ElementsBegin()->GetGeometry();
+        ElementSizeFunctionType minimum_h_func = FluidCharacteristicNumbersUtilities::GetMinimumElementSizeFunction(r_geom);
+
+        // Set the function to calculate the Fourier numbers
+        std::function<double(const Element&, const ElementSizeFunctionType&, const double)> thermal_fourier_number_function;
+        if (mConsiderArtificialDiffusion) {
+            if (mNodalDensityFormulation) {
+                thermal_fourier_number_function = FluidCharacteristicNumbersUtilities::CalculateElementThermalFourierNumber<true,true>;
+            } else {
+                thermal_fourier_number_function = FluidCharacteristicNumbersUtilities::CalculateElementThermalFourierNumber<true,false>;
+            }
+        } else {
+            if (mNodalDensityFormulation) {
+                thermal_fourier_number_function = FluidCharacteristicNumbersUtilities::CalculateElementThermalFourierNumber<false,true>;
+            } else {
+                thermal_fourier_number_function = FluidCharacteristicNumbersUtilities::CalculateElementThermalFourierNumber<false,false>;
+            }
+        }
+
+        // Obtain the maximum CFL and Peclet numbers
+        double max_CFL, max_Fo_k;
+        const double current_dt = mrModelPart.GetProcessInfo().GetValue(DELTA_TIME);
+        typedef CombinedReduction<MaxReduction<double>, MaxReduction<double>> CombinedMaxReduction;
+        std::tie(max_CFL, max_Fo_k) = block_for_each<CombinedMaxReduction>(mrModelPart.Elements(), [&](Element& rElement){
+            const double CFL = FluidCharacteristicNumbersUtilities::CalculateElementCFL(rElement, minimum_h_func, current_dt);
+            const double thermal_Fo_number = thermal_fourier_number_function(rElement, minimum_h_func, current_dt);
+            return std::make_tuple(CFL, thermal_Fo_number);
+        });
+
+        // Calculate the new time increment from the maximum characteristic numbers in the mesh
+        const double new_dt = CalculateNewDeltaTime(
+            current_dt,
+            std::make_pair(max_CFL, mCFL),
+            std::make_pair(max_Fo_k, mThermalFourier));
+
+        return new_dt;
+
+        KRATOS_CATCH("")
+    }
+
+    template<>
+    double EstimateDtUtility::InternalEstimateDt<true,true,true>() const
+    {
+        KRATOS_TRY;
+
+        // Get the minimum element size function according to the corresponding geometry for the CFL calculation
+        // Note that in here it is assumed that all the elements in the model part feature the same geometry
+        const auto& r_geom = mrModelPart.ElementsBegin()->GetGeometry();
+        ElementSizeFunctionType minimum_h_func = FluidCharacteristicNumbersUtilities::GetMinimumElementSizeFunction(r_geom);
+
+        // Set the function to calculate the Fourier numbers
+        std::function<std::tuple<double,double>(const Element&, const ElementSizeFunctionType&, const double)> fourier_numbers_function;
```

Comment:
wouldn't it be easier with a pair instead of a tuple?

Reply:
Yes. We can use a pair in this case. However, this would imply changing the characteristic number utils. If you don't mind doing it in this PR I'll proceed with the changes.

Reply:
I changed all of them except the one in the custom reduction. If we use an `std::pair` in there I got a compilation error when unpacking the result of the reduction :confused: